feat: Add support for Gitea delete and pull_request_comment Gitea webhook events (#34224)

Aneesh-Hegde opened the pull request:
This pull request introduces support for the 'delete' and 'pull_request_comment' webhook events from Gitea to the Zulip integration.

This PR enhances the Gitea webhook integration by adding functionality to handle notifications for the following events:

- `delete`: Triggered when branches or tags are deleted in a Gitea repository.
- `pull_request_comment`: Triggered when comments are added or edited on pull requests in Gitea.

This implementation includes:

- Handling of the webhook payloads for these events in `zerver/webhooks/gogs/view.py`.
- Creation of appropriate Zulip message content based on the event type.
- Addition of test cases in `zerver/webhooks/gogs/tests.py` to ensure correct functionality.

Fixes: #34224 Gitea webhook integration - Improve event coverage

**Screenshots and screen captures:**

<img width="286" alt="Screenshot 2025-03-30 at 1 18 16 PM" src="https://github.com/user-attachments/assets/dae1f932-afcb-440f-b91f-b9796009cf02" />

* **Branch Deletion:** A screenshot showing the Zulip message when a branch is deleted in Gitea.
<img width="755" alt="Screenshot 2025-03-30 at 1 18 43 PM" src="https://github.com/user-attachments/assets/45122c95-36f7-4526-9e2a-2a7aa0c2240d" />

* **Tag Deletion:** A screenshot showing the Zulip message when a tag is deleted in Gitea.
<img width="755" alt="Screenshot 2025-03-30 at 1 18 38 PM" src="https://github.com/user-attachments/assets/bebe0723-7cfc-4fad-b913-9670c530c539" />

* **Pull Request Comment Created:** A screenshot showing the Zulip message when a new comment is added to a pull request in Gitea.
* **Pull Request Comment Edited:** A screenshot showing the Zulip message when an existing comment on a pull request is edited in Gitea.
<img width="755" alt="Screenshot 2025-03-30 at 1 18 31 PM" src="https://github.com/user-attachments/assets/79f8c641-5967-4295-9dba-0750b6af9bf5" />

<details>
<summary>Self-review checklist</summary>

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
        (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Niloth-p reviewed (commented):
Thank you for working on this.
I've done an initial review.
Could you please split this into 2 commits, one for each event type - "pull_request_comment" and "delete"?
Thanks for adding support for the deletion of tags as well.

Aneesh-Hegde commented on zerver/webhooks/gitea/tests.py:
```diff
@@ -108,3 +108,42 @@ def test_push_filtered_by_branches_ignore(
+            "issue_comment__edited_pr",
+            expected_topic_name,
+            expected_message,
+            HTTP_X_GITEA_EVENT_TYPE="pull_request_comment",
```
The issue_comment header has two types: issue_comment, which is triggered when someone comments on an issue, and pull_request_comment, which is triggered when someone comments on the pull request. So we need the issue_comment as the header, which is extracted from `fixture_to_headers` in `zerver/lib/webhooks/common.py`. This is why I named it like that. Please let me know if I have misunderstood the problem statement.
<img width="599" alt="Screenshot 2025-03-31 at 3 03 48 PM" src="https://github.com/user-attachments/assets/f2d0963e-52d8-4860-99f6-fc6649c03be8" />

Aneesh-Hegde commented on zerver/webhooks/gogs/view.py:
```diff
@@ -198,6 +218,14 @@ def gogs_webhook_main(
 ) -> HttpResponse:
     repo = payload["repository"]["name"].tame(check_string)
     event = validate_extract_webhook_http_header(request, http_header_name, integration_name)
+    try:
```
Should I make this changes in `gitea/view.py` as previous implementation of gitea webhook was done on this file because gitea is build upon gogs so implementation of both webhook is almost same.

Aneesh-Hegde commented on zerver/webhooks/gogs/view.py:
```diff
@@ -236,15 +264,40 @@ def gogs_webhook_main(
+        issue_title = payload["issue"]["title"].tame(check_string)
+        repo_name = payload["repository"]["name"].tame(check_string)
+
+        if event_type == "pull_request_comment":
```
[same justification](https://github.com/zulip/zulip/pull/34230/commits/847c2266f1347d7a663649d32085de7144f53d94#r2020768998)

Aneesh-Hegde commented on zerver/webhooks/gogs/view.py:
```diff
@@ -236,15 +264,40 @@ def gogs_webhook_main(
-            type="issue",
-            id=payload["issue"]["number"].tame(check_int),
-            title=payload["issue"]["title"].tame(check_string),
+            repo=repo_name,
```
I thought that extracting issue_number, issue_title, and repo_name improves readability and is a small refactor.  Would you like me to move the variable extraction to a separate commit or use the previous implementation ?

Aneesh-Hegde commented on zerver/webhooks/gogs/view.py:
```diff
@@ -256,6 +309,25 @@ def gogs_webhook_main(
             title=payload["release"]["name"].tame(check_string),
         )
+    elif event == "delete":
+        if payload["ref_type"].tame(check_string) == "branch":
```
Yes we can do similar to implementation done in `github/view.py` using `EVENT_FUNCTION_MAPPER`.

Aneesh-Hegde commented on zerver/webhooks/gitea/fixtures/delete__branch.json:
```diff
@@ -0,0 +1,124 @@
+{
```
The only difference is the header that is passed.Since we need delete header from `fixture_to_headers` function so I created a similar file and gave name as `delete__branch.json`.

Aneesh-Hegde committed:
gogs: Replace if-elif chain with event handler registry.

Use dictionary mapping for Gogs/Gitea webhook events instead of
lengthy if-elif chains. This improves maintainability and makes
adding new event types simpler.

The refactoring introduces handler functions for each event type
while maintaining identical functionality.

Aneesh-Hegde commented:
@Niloth-p Will you take a look at PR I have made the necessary changes

Aneesh-Hegde committed:
webhooks: Add delete event support for Gogs/Gitea.

Implement branch and tag delete event handlers. Adds handle_delete_event
function, message templates, and test cases for both event types.

Notifications show the actor and deleted reference name when users
delete branches or tags.

Aneesh-Hegde committed:
gitea: Add pull request comment support.
Implement PR comment event handling for created, edited, and deleted
comments. Adds handler function, message template, fixtures, and
comprehensive test coverage.

Aneesh-Hegde committed:
gogs: Add fixtures for delete events.

Add test fixtures for branch and tag delete events from Gogs/Gitea.
Includes delete__branch.json and delete__tag.json with complete
payload structures for testing delete event handlers.

Aneesh-Hegde committed:
gogs: Add fixture for PR comment events.
Add fixture for testing pull request
comment event handling in Gogs/Gitea webhooks.

Aneesh-Hegde commented:
@timabbott @Niloth-p Gentle ping

Aneesh-Hegde commented:
Hey @Niloth-p PTAL.
CC: @alya
